pkg/integration/transaction_test.go

```diff
@@ -0,0 +1,234 @@
+	options := server.DefaultOptions()
+	bs := servertest.NewBufconnServer(options)
+
+	defer os.RemoveAll(options.Dir)
+	defer os.Remove(".state-")
+
+	bs.Start()
+	defer bs.Stop()
+
+	client := ic.DefaultClient().WithOptions(ic.DefaultOptions().WithDialOptions([]grpc.DialOption{grpc.WithContextDialer(bs.Dialer), grpc.WithInsecure()}))
+
+	serverUUID, sessionID, err := client.OpenSession(context.TODO(), []byte(`immudb`), []byte(`immudb`), "defaultdb")
+	require.NoError(t, err)
+	require.NotNil(t, serverUUID)
+	require.NotNil(t, sessionID)
+
+	tx1, err := client.BeginTx(context.TODO(), &ic.TxOptions{TxMode: schema.TxMode_READ_WRITE})
+	require.NoError(t, err)
+	err = tx1.TxSQLExec(context.TODO(), `CREATE TABLE table1(id INTEGER,PRIMARY KEY id);`, nil)
+	_, err = tx1.Commit(context.TODO())
+	require.NoError(t, err)
+
+	tx1, err = client.BeginTx(context.TODO(), &ic.TxOptions{TxMode: schema.TxMode_READ_WRITE})
+	require.NoError(t, err)
+	err = tx1.TxSQLExec(context.TODO(), `CREATE TABLE table2(id INTEGER,PRIMARY KEY id);`, nil)
+	require.NoError(t, err)
+
+	tx2, err := client.BeginTx(context.TODO(), &ic.TxOptions{TxMode: schema.TxMode_READ_ONLY})
+	err = tx2.TxSQLExec(context.TODO(), `CREATE TABLE table1(id INTEGER,PRIMARY KEY id);`, nil)
+	require.Error(t, err)
+	require.Equal(t, err.(immuErrors.ImmuError).Error(), "read write transaction not ongoing")
+	_, err = tx2.TxSQLQuery(context.TODO(), "SELECT * FROM table1;", nil, true)
+	require.NoError(t, err)
+
+	tx3, err := client.BeginTx(context.TODO(), &ic.TxOptions{TxMode: schema.TxMode_READ_ONLY})
+	require.NoError(t, err)
+	_, err = tx3.TxSQLQuery(context.TODO(), "SELECT * FROM table1;", nil, true)
+	require.NoError(t, err)
+}
+
+func TestTransaction_ChangingDBOnSessionNoError(t *testing.T) {
+	options := server.DefaultOptions()
+	bs := servertest.NewBufconnServer(options)
+
+	defer os.RemoveAll(options.Dir)
+	defer os.Remove(".state-")
+
+	bs.Start()
+	defer bs.Stop()
+
+	client := ic.DefaultClient().WithOptions(ic.DefaultOptions().WithDialOptions([]grpc.DialOption{grpc.WithContextDialer(bs.Dialer), grpc.WithInsecure()}))
+
+	_, _, err := client.OpenSession(context.TODO(), []byte(`immudb`), []byte(`immudb`), "defaultdb")
+	require.NoError(t, err)
+	txDefaultDB, err := client.BeginTx(context.TODO(), &ic.TxOptions{TxMode: schema.TxMode_READ_WRITE})
+	err = txDefaultDB.TxSQLExec(context.TODO(), `CREATE TABLE tableDefaultDB(id INTEGER,PRIMARY KEY id);`, nil)
+	require.NoError(t, err)
+
+	client2 := ic.DefaultClient().WithOptions(ic.DefaultOptions().WithDialOptions([]grpc.DialOption{grpc.WithContextDialer(bs.Dialer), grpc.WithInsecure()}))
+	_, _, err = client2.OpenSession(context.TODO(), []byte(`immudb`), []byte(`immudb`), "defaultdb")
+	require.NoError(t, err)
+	err = client2.CreateDatabase(context.TODO(), &schema.DatabaseSettings{DatabaseName: "db2"})
+	require.NoError(t, err)
+	_, err = client2.UseDatabase(context.TODO(), &schema.Database{DatabaseName: "db2"})
```

Comment:
Note: `UseDatabase` should be as thread-safe as immuclient, it may need to acquire a lock

Reply:
I removed mutex on immuclient for the moment